runtime/parachains/src/paras.rs

```diff
@@ -0,0 +1,1103 @@
+						.take_while(|&(_, at)| at <= &pruning_height)
+						.count();
+					(up_to_idx, pruning_tasks.drain(..up_to_idx))
+				};
+
+				for (para_id, _) in pruning_tasks_to_do {
+					let full_deactivate = <Self as Store>::PastCodeMeta::mutate(&para_id, |meta| {
+						for pruned_repl_at in meta.prune_up_to(pruning_height) {
+							<Self as Store>::PastCode::remove(&(para_id, pruned_repl_at));
+						}
+
+						meta.most_recent_change().is_none() && Self::parachain_head(&para_id).is_none()
+					});
+
+					// This parachain has been removed and now the vestigial code
+					// has been removed from the state. clean up meta as well.
+					if full_deactivate {
+						<Self as Store>::PastCodeMeta::remove(&para_id);
+					}
+				}
+
+				pruning_tasks_done as u64
+			});
+
+		// 1 read for the meta for each pruning task, 1 read for the config
+		// 2 writes: updating the meta and pruning the code
+		T::DbWeight::get().reads_writes(1 + pruning_tasks_done, 2 * pruning_tasks_done)
+	}
+
+	/// Schedule a para to be initialized at the start of the next session.
+	#[allow(unused)]
+	pub(crate) fn schedule_para_initialize(id: ParaId, genesis: ParaGenesisArgs) -> Weight {
+		let dup = UpcomingParas::mutate(|v| {
+			match v.binary_search(&id) {
+				Ok(_) => true,
+				Err(i) => {
+					v.insert(i, id);
+					false
+				}
+			}
+		});
+
+		if dup {
+			let weight = T::DbWeight::get().reads_writes(1, 0);
+			return weight;
+		}
+
+		UpcomingParasGenesis::insert(&id, &genesis);
+
+		T::DbWeight::get().reads_writes(1, 2)
+	}
+
+	/// Schedule a para to be cleaned up at the start of the next session.
+	#[allow(unused)]
+	pub(crate) fn schedule_para_cleanup(id: ParaId) -> Weight {
+		OutgoingParas::mutate(|v| {
+			match v.binary_search(&id) {
+				Ok(_) => T::DbWeight::get().reads_writes(1, 0),
+				Err(i) => {
+					v.insert(i, id);
+					T::DbWeight::get().reads_writes(1, 1)
+				}
+			}
+		})
+	}
+
+	/// Schedule a future code upgrade of the given parachain, to be applied after inclusion
+	/// of a block of the same parachain executed in the context of a relay-chain block
+	/// with number >= `expected_at`
+	///
+	/// If there is already a scheduled code upgrade for the para, this is a no-op.
+	#[allow(unused)]
+	pub(crate) fn schedule_code_upgrade(
+		id: ParaId,
+		new_code: ValidationCode,
+		expected_at: T::BlockNumber,
+	) -> Weight {
+		<Self as Store>::FutureCodeUpgrades::mutate(&id, |up| {
+			if up.is_some() {
+				T::DbWeight::get().reads_writes(1, 0)
+			} else {
+				*up = Some(expected_at);
+				FutureCode::insert(&id, new_code);
+				T::DbWeight::get().reads_writes(1, 2)
+			}
+		})
+	}
+
+	/// Note that a para has progressed to a new head, where the new head was executed in the context
+	/// of a relay-chain block with given number. This will apply pending code upgrades based
+	/// on the block number provided.
+	#[allow(unused)]
+	pub(crate) fn note_new_head(
+		id: ParaId,
+		new_head: HeadData,
+		execution_context: T::BlockNumber,
```

Comment:
`context` suggests having a inherent state, but this really is just a `execution_at`, like a time instant, isn't it?

Reply:
In which cases does `execution_context` differ from `now`?